fix(withTbleSettings): isSelected -> selected

GermanVor opened the pull request.

amje commented on src/components/Table/README.md:
```diff
@@ -263,7 +263,7 @@ const MyTable1 = withTableSettings({sortable: false})(Table);
 type TableSettingsData = Array<{
   id: string;
-  isSelected?: boolean;
+  selected?: boolean;
```
Why this change? It's a breaking change

teleginzhenya replied:
agree, need to separate `TableColumnSetupItem` from `TableSettingsData` and return backward compatibility for `TableColumnSetupItem` only

amje reviewed (changes requested):
It breaks current behaviour

teleginzhenya commented on src/components/Table/hoc/withTableSettings/withTableSettings.tsx:
```diff
@@ -23,26 +23,26 @@ import './withTableSettings.scss';
 
 export type TableSettingsData = TableSetting[];
 
 export type TableColumnSetupItem = TableSetting & {
```
```suggestion
export type TableColumnSetupItem = {
```

teleginzhenya commented on src/components/Table/hoc/withTableSettings/withTableSettings.tsx:
```diff
@@ -23,26 +23,26 @@ import './withTableSettings.scss';
 export type TableColumnSetupItem = TableSetting & {
     title: React.ReactNode;
-    isRequired?: boolean;
+    required?: boolean;
 };
```
```suggestion
    selected?: boolean;
    required?: boolean;
};
```

teleginzhenya commented on src/components/Table/hoc/withTableSettings/withTableSettings.tsx:
```diff
@@ -23,26 +23,26 @@ import './withTableSettings.scss';
 export type TableSetting = {
     id: string;
-    isSelected?: boolean;
+    selected?: boolean;
```
```suggestion
    isSelected?: boolean;
```

GermanVor commented:
The problem is real

I rewrote [TableColumnSetup](https://github.com/gravity-ui/uikit/blob/main/src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx#L198) but I did not think that it was [public component](https://github.com/gravity-ui/uikit/blob/main/src/components/Table/hoc/index.ts#L25). Interfaces `TableColumnSetupItem` also is public, but it is not connected with the [TableSetting](https://github.com/gravity-ui/uikit/blob/main/src/components/Table/README.md#tablesettingsdata).

So, the best solution now to save backward compatibility in my opinion 
 - stop exporting new  TableColumnSetup
 - create separated TableColumnSetup component with code before I rewrote TableColumnSetup and export it
 
 3616e97c5a1618b6ab8e1e8a0f0ddded1dad4b87
 645df5615cc52f419cd7cb57865af454ba1ad8be
 
 What to do with new TableColumnSetup component I do not know - probably delete it in future, but now save it for backward compatibility.
 
 @teleginzhenya @Raubzeug

GermanVor commented:
@amje @korvin89 @Raubzeug @ValeraS 

I begin rewrite new `TableColumnSetup` with the props of old one and I have new thoughts:

- If I rewrite a new component (`TableColumnSetup` with `TreeSelect`) for the old ones, then I risk breaking the new component and the old component (`TableColumnSetup` with `List`).

- The old component had props, which it is unclear how to apply to the new component (for example `filter` - there is no such property in `TreeSelect`)

- The behavior of the old component is slightly different from the new one - in the new component we allow the required columns to be moved, but we do not allow to hide them, while in old component case we moved `required` columns at the start of the list, without opportunity move them.

- The old component had the className option - it is not difficult to return it, but if the users used complex selectors, they are very likely will be broken cause `TreeSelect` probably is not comparable with `List` and `Popup`.

The best way to save  backward compatibility in my opinion will be create `TableColumnSetup` with old code (code before i rewrote `TableColumnSetup` with `TreeSelect`) and export it instead of `TableColumnSetup` from `withTableSettings`. 

In this case we 100% sure that new component will not be broken, users that use `TableColumnSetup` will get back component that they expected, old component 100% will not be broken.

And after that we will think what to do with `TableColumnSetup`.

Raubzeug commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -18,7 +18,7 @@ import type {
 import {ListContainerView, ListItemView} from '../../../../useList';
 import {block} from '../../../../utils/cn';
-import type {TableColumnSetupItem, TableSetting} from '../withTableSettings';
+import type {TableSetting} from '../withTableSettings';
```
We have a circular dependency here, let's avoid such a case.

Raubzeug commented on src/components/TableColumnSetup/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './TableColumnSetup';
```
Should export it also in https://github.com/gravity-ui/uikit/blob/645df5615cc52f419cd7cb57865af454ba1ad8be/src/components/index.ts

GermanVor replied:
bcde4ae403818d7103a6bac8874b87ba08d2f7da

GermanVor commented on src/components/Table/__tests__/Table.withTableSettings.test.tsx:
```diff
@@ -288,7 +288,7 @@ describe('withTableSettings', () => {
             await userEvent.click(screen.getByRole('button', {name: 'Table settings'}));
             await userEvent.click(await screen.findByRole('button', {name: 'description'}));
-            await userEvent.click(screen.getByRole('button', {name: 'Apply'}));
+            await userEvent.click(screen.getByRole('button', {name: 'button_apply'}));
```
have no idea why it is broken now

Raubzeug replied:
This is because you test component `const TableWithSettings = withTableSettings<SomeItem>(Table);` . And keysets are taken from https://github.com/gravity-ui/uikit/tree/839d503f378e23eef87887866d20973bcd174fb4/src/components/Table/hoc/withTableSettings/i18n

GermanVor replied:
In separate PR, I do not know what happened with this place

GermanVor commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -46,17 +51,63 @@ const reorderArray = <T extends unknown>(list: T[], startIndex: number, endIndex
+    let lastStickyLeftIdx = 0;
+    for (; lastStickyLeftIdx !== visibleFlattenIds.length; lastStickyLeftIdx++) {
+        const visibleFlattenId = visibleFlattenIds[lastStickyLeftIdx];
+        const item = items.find((item) => item.id === visibleFlattenId) as Item | undefined;
```
@IsaevAlexandr `ListItemType` generic works bad. 
Do not know why.

GermanVor replied:
459739976c98c1efada3d67a9d3f81c468ac382d

GermanVor commented on src/components/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -0,0 +1,124 @@
+        title,
+        isRequired: required,
+        isSelected: selected,
+    }));
```
839d503f378e23eef87887866d20973bcd174fb4

GermanVor commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -46,17 +51,63 @@ const reorderArray = <T extends unknown>(list: T[], startIndex: number, endIndex
+    for (let i = firstStickyRightIdx; i < visibleFlattenIds.length; i++) {
+        const visibleFlattenId = visibleFlattenIds[i];
+        stickyRightItemIdList.push(visibleFlattenId);
+    }
```
a2afc9045009e232fd1b4f3a3869c7a6167bedc8

GermanVor commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -46,17 +51,63 @@ const reorderArray = <T extends unknown>(list: T[], startIndex: number, endIndex
 };
 
+const prepareStikyState = (items: ListItemType<Item>[], visibleFlattenIds: string[]) => {
+    let lastStickyLeftIdx = 0;
```
10822427d165c0a1f07184a226357379b5d86c0a

GermanVor commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -32,11 +34,14 @@ function identity<T>(value: T): T {
 const controlsCn = b('controls');
 
-const reorderArray = <T extends unknown>(list: T[], startIndex: number, endIndex: number): T[] => {
+const reorderArray = <T extends TableColumnSetupItem>(
```
1876c01d32f171627da815cc5d5b7c9c5cbbe130

GermanVor commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -302,7 +365,7 @@ export const TableColumnSetup = (props: TableColumnSetupProps) => {
         <TreeSelect
-            className={tableColumnSetupCn}
+            className={b(null, className)}
             mapItemDataToProps={identity}
```
I think main reason is [here](https://github.com/gravity-ui/uikit/blob/6aa5de98d89e0b82b7b9f357dc1dbc6bde0bd37b/src/components/useList/components/ListItemView/ListItemView.tsx#L113)

GermanVor replied:
6aa5de98d89e0b82b7b9f357dc1dbc6bde0bd37b

ValeraS commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -201,9 +201,11 @@ const useDndRenderItem = (sortable: boolean | undefined) => {
+        const commonProps: ListItemViewProps = {
             ...props,
-            ...data,
+            ...itemViewPropsFromData,
```
Why is it here? all needed data already in props via `mapItemDataToProps`

ValeraS commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -365,7 +378,7 @@ export const TableColumnSetup = (props: TableColumnSetupProps) => {
         <TreeSelect
             className={b(null, className)}
-            mapItemDataToProps={identity}
+            mapItemDataToProps={mapItemDataToProps}
```
Why is `dndItems` needed? `mapItemDataToProps` translates items to props, no need to prepare items for that.

amje commented on src/components/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -0,0 +1,130 @@
+import React from 'react';
```
Why we can't just move the component outside the Table?

amje replied:
Let's simply move it up to the components. It should be the one component. Keep backward compatibility with old versions and add sticky feature

GermanVor commented on src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:
```diff
@@ -44,17 +41,35 @@ const reorderArray = <T extends unknown>(list: T[], startIndex: number, endIndex
-const prepareDndItems = (tableColumnItems: TableColumnSetupItem[]) => {
-    return tableColumnItems.map<Item>((tableColumnItem) => {
-        const hasSelectionIcon = tableColumnItem.isRequired === false;
+const prepareStikyState = (
```
79d33de373c7243a52456037e6c3b4c557884821
